chain/deals/client_states.go

```diff
@@ -1,142 +1,224 @@
+func (c *Client) staged(ctx context.Context, deal ClientDeal) (func(*ClientDeal), error) {
+	// TODO: Maybe wait for pre-commit
 
-	resp, err := c.readStorageDealResp(deal)
-	if err != nil {
-		return err
-	}
+	return nil, nil
+}
 
-	if resp.State != api.DealStaged {
-		return xerrors.Errorf("deal wasn't staged (State=%d)", resp.State)
-	}
+func (c *Client) sealing(ctx context.Context, deal ClientDeal) (func(*ClientDeal), error) {
+	checkFunc := func(ts *types.TipSet) (done bool, more bool, err error) {
+		sd, err := stmgr.GetStorageDeal(ctx, c.sm, deal.DealID, ts)
+		if err != nil {
+			// TODO: This may be fine for some errors
+			return false, false, xerrors.Errorf("failed to look up deal on chain: %w", err)
+		}
 
-	log.Info("DEAL STAGED!")
+		if sd.ActivationEpoch > 0 {
+			select {
+			case c.updated <- clientDealUpdate{
+				newState: api.DealComplete,
+				id:       deal.ProposalCid,
+			}:
+			case <-c.stop:
+			}
 
-	return nil
-}
+			return true, false, nil
+		}
 
-func (c *Client) staged(ctx context.Context, deal ClientDeal) error {
-	/* miner seals our data, hopefully */
+		return false, true, nil
+	}
+
+	called := func(msg *types.Message, ts *types.TipSet, curH uint64) (more bool, err error) {
+		defer func() {
+			if err != nil {
+				select {
+				case c.updated <- clientDealUpdate{
+					newState: api.DealComplete,
+					id:       deal.ProposalCid,
+					err:      xerrors.Errorf("handling applied event: %w", err),
+				}:
+				case <-c.stop:
+				}
+			}
+		}()
+
+		if msg == nil {
+			log.Error("timed out waiting for deal activation... what now?")
+			return false, nil
+		}
 
-	resp, err := c.readStorageDealResp(deal)
-	if err != nil {
-		return err
-	}
+		var params actors.SectorProveCommitInfo
+		if err := params.UnmarshalCBOR(bytes.NewReader(msg.Params)); err != nil {
+			return false, err
+		}
 
-	if resp.State != api.DealSealing {
-		return xerrors.Errorf("deal wasn't sealed (State=%d)", resp.State)
-	}
+		var found bool
+		for _, dealID := range params.DealIDs {
+			if dealID == deal.DealID {
+				found = true
+				break
+			}
+		}
 
-	log.Info("DEAL SEALED!")
+		if !found {
+			return true, nil
+		}
 
-	// TODO: want?
-	/*ssize, err := stmgr.GetMinerSectorSize(ctx, c.sm, nil, deal.Proposal.MinerAddress)
-	if err != nil {
-		return xerrors.Errorf("failed to get miner sector size: %w", err)
-	}
+		sd, err := stmgr.GetStorageDeal(ctx, c.sm, deal.DealID, ts)
+		if err != nil {
+			return false, xerrors.Errorf("failed to look up deal on chain: %w", err)
+		}
 
-	ok, err := sectorbuilder.VerifyPieceInclusionProof(ssize, deal.Proposal.Size, deal.Proposal.CommP, resp.CommD, resp.PieceInclusionProof.ProofElements)
-	if err != nil {
-		return xerrors.Errorf("verifying piece inclusion proof in staged deal %s: %w", deal.ProposalCid, err)
-	}
-	if !ok {
-		return xerrors.Errorf("verifying piece inclusion proof in staged deal %s failed", deal.ProposalCid)
-	}*/
+		if sd.ActivationEpoch == 0 {
+			return true, nil
```

Comment:
how would this happen? If we found the deal in the message, but it ends up not getting activated... maybe this is an error?